In [1]:
import sys
print(sys.executable)


/Users/sai/Desktop/luxury_housing/.venv/bin/python


In [1]:
import pandas as pd
import os

BASE_DIR = os.getcwd()
CSV_PATH = os.path.join(
    BASE_DIR,
    "..",
    "01_raw_data",
    "Luxury_Housing_Bangalore.csv"
)

df = pd.read_csv(CSV_PATH)

df.head()


Property_ID       Micro_Market Project_Name Developer_Name  Unit_Size_Sqft  \
0  PROP000001      Sarjapur Road    Project_0            RMZ          4025.0   
1  PROP000002        Indiranagar    Project_1    Puravankara          5760.0   
2  PROP000003  Bannerghatta Road    Project_2   Tata Housing          7707.0   
3  PROP000004       bellary road    Project_3        Embassy          6192.0   
4  PROP000005        Koramangala    Project_4        SNN Raj          7147.0   

  Configuration     Ticket_Price_Cr Transaction_Type Buyer_Type  \
0          4bhk  12.750846039118798          Primary        NRI   
1          3Bhk  16.292151871065954          Primary      Other   
2          4bhk  10.517724412961911          Primary        HNI   
3          3BHK   9.396367494232896          Primary        HNI   
4          4Bhk  15.345392444511946        Secondary        HNI   

  Purchase_Quarter  Connectivity_Score  Amenity_Score   Possession_Status  \
0       2025-03-31            7.990091       5.462863              Launch   
1       2024-06-30            4.839024            NaN  Under construction   
2       2023-12-31            8.131315       8.669227       Ready to move   
3       2024-03-31            7.501657       5.720246       Ready to move   
4       2024-12-31            4.525216       8.609649  Under construction   

  Sales_Channel NRI_Buyer  Locality_Infra_Score  Avg_Traffic_Time_Min  \
0        Broker       yes              9.212491                    18   
1      NRI Desk        no              7.723898                   106   
2        Direct       yes              6.985493                   113   
3        Online       yes              6.100929                   106   
4        Broker        no              5.312510                    18   

              Buyer_Comments  
0       Loved the amenities!  
1                        NaN  
2  Agent was not responsive.  
3        Excellent location!  
4    Too far from my office.

In [2]:
# Row count and column count
rows, cols = df.shape

print("Total Rows:", rows)
print("Total Columns:", cols)


Total Rows: 101000
Total Columns: 18


In [3]:
# Column names and data types
df.dtypes


Property_ID                 str
Micro_Market                str
Project_Name                str
Developer_Name              str
Unit_Size_Sqft          float64
Configuration               str
Ticket_Price_Cr             str
Transaction_Type            str
Buyer_Type                  str
Purchase_Quarter            str
Connectivity_Score      float64
Amenity_Score           float64
Possession_Status           str
Sales_Channel               str
NRI_Buyer                   str
Locality_Infra_Score    float64
Avg_Traffic_Time_Min      int64
Buyer_Comments              str
dtype: object

In [4]:
# Column-wise non-null count
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 101000 entries, 0 to 100999
Data columns (total 18 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Property_ID           101000 non-null  str    
 1   Micro_Market          101000 non-null  str    
 2   Project_Name          101000 non-null  str    
 3   Developer_Name        101000 non-null  str    
 4   Unit_Size_Sqft        90954 non-null   float64
 5   Configuration         101000 non-null  str    
 6   Ticket_Price_Cr       90981 non-null   str    
 7   Transaction_Type      101000 non-null  str    
 8   Buyer_Type            101000 non-null  str    
 9   Purchase_Quarter      101000 non-null  str    
 10  Connectivity_Score    101000 non-null  float64
 11  Amenity_Score         90910 non-null   float64
 12  Possession_Status     101000 non-null  str    
 13  Sales_Channel         101000 non-null  str    
 14  NRI_Buyer             101000 non-null  str    
 15  Locality_In

In [5]:
# Missing values per column
missing_counts = df.isnull().sum()
missing_counts


Property_ID                 0
Micro_Market                0
Project_Name                0
Developer_Name              0
Unit_Size_Sqft          10046
Configuration               0
Ticket_Price_Cr         10019
Transaction_Type            0
Buyer_Type                  0
Purchase_Quarter            0
Connectivity_Score          0
Amenity_Score           10090
Possession_Status           0
Sales_Channel               0
NRI_Buyer                   0
Locality_Infra_Score        0
Avg_Traffic_Time_Min        0
Buyer_Comments          18287
dtype: int64

In [6]:
# Missing value percentage per column
missing_percent = (df.isnull().sum() / len(df)) * 100
missing_percent.round(2)


Property_ID              0.00
Micro_Market             0.00
Project_Name             0.00
Developer_Name           0.00
Unit_Size_Sqft           9.95
Configuration            0.00
Ticket_Price_Cr          9.92
Transaction_Type         0.00
Buyer_Type               0.00
Purchase_Quarter         0.00
Connectivity_Score       0.00
Amenity_Score            9.99
Possession_Status        0.00
Sales_Channel            0.00
NRI_Buyer                0.00
Locality_Infra_Score     0.00
Avg_Traffic_Time_Min     0.00
Buyer_Comments          18.11
dtype: float64

In [7]:
# Columns with missing values only
missing_summary = pd.concat(
    [missing_counts, missing_percent],
    axis=1,
    keys=["Missing_Count", "Missing_Percentage"]
)

missing_summary[missing_summary["Missing_Count"] > 0].sort_values(
    by="Missing_Percentage",
    ascending=False
)


Missing_Count  Missing_Percentage
Buyer_Comments           18287           18.105941
Amenity_Score            10090            9.990099
Unit_Size_Sqft           10046            9.946535
Ticket_Price_Cr          10019            9.919802

In [8]:
# Duplicate record check
duplicate_rows = df.duplicated().sum()
duplicate_rows


np.int64(1000)

In [9]:
# Duplicate Property_ID check
df['Property_ID'].duplicated().sum()


np.int64(1000)

In [10]:
# Basic summary statistics (numeric columns)
df.describe()


Unit_Size_Sqft  Connectivity_Score  Amenity_Score  \
count    90954.000000       101000.000000   90910.000000   
mean      5972.274765            6.992619       7.503663   
std       1783.397836            1.731757       1.440758   
min         -1.000000            4.000031       5.000224   
25%       4477.250000            5.494526       6.260329   
50%       5990.000000            6.985805       7.499123   
75%       7497.000000            8.490000       8.752207   
max       8999.000000            9.999970       9.999865   

       Locality_Infra_Score  Avg_Traffic_Time_Min  
count         101000.000000         101000.000000  
mean               7.498426             67.182921  
std                1.443128             30.271611  
min                5.000013             15.000000  
25%                6.247954             41.000000  
50%                7.495614             67.000000  
75%                8.749824             93.000000  
max                9.999956            119.000000

In [11]:
# Categorical column overview
df.describe(include='object')


/var/folders/93/71dbkqd911dfzn8p55s7072m0000gn/T/ipykernel_14642/2406388679.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include='object')


Property_ID Micro_Market Project_Name Developer_Name Configuration  \
count       101000       101000       101000         101000        101000   
unique      100000           48          500             11             9   
top     PROP000051    Jayanagar    Project_7       Prestige         5Bhk+   
freq             2         2176          207           9318         11375   

       Ticket_Price_Cr Transaction_Type Buyer_Type Purchase_Quarter  \
count            90981           101000     101000           101000   
unique           72854                2          5                8   
top              100.0        Secondary        CXO       2024-03-31   
freq               278            50506      20340            12750   

         Possession_Status Sales_Channel NRI_Buyer          Buyer_Comments  
count               101000        101000    101000                   82713  
unique                   3             4         2                       9  
top     Under construction        Online        no  Great value for money.  
freq                 33769         25411     50516                    9383

In [12]:
# Micro_Market distribution
df['Micro_Market'].value_counts().head(10)

Micro_Market
Jayanagar            2176
Jp Nagar             2167
Bannerghatta Road    2164
jp nagar             2163
mg road              2160
domlur               2156
sarjapur road        2152
indiranagar          2149
bannerghatta road    2149
SARJAPUR ROAD        2149
Name: count, dtype: int64

In [13]:
# Developer distribution
df['Developer_Name'].value_counts().head(10)


Developer_Name
Prestige             9318
SNN Raj              9309
Brigade              9240
L&T Realty           9225
Godrej               9215
Total Environment    9209
Puravankara          9166
Embassy              9164
RMZ                  9092
Sobha                9043
Name: count, dtype: int64

In [14]:
# Configuration distribution
df['Configuration'].value_counts()

Configuration
5Bhk+    11375
3Bhk     11332
5BHK+    11291
3bhk     11262
5bhk+    11210
4bhk     11173
4Bhk     11151
3BHK     11136
4BHK     11070
Name: count, dtype: int64

In [15]:
# Transaction type distribution
df['Transaction_Type'].value_counts()

Transaction_Type
Secondary    50506
Primary      50494
Name: count, dtype: int64

In [16]:
# Buyer type distribution
df['Buyer_Type'].value_counts()


Buyer_Type
CXO                20340
NRI                20207
Startup Founder    20185
Other              20149
HNI                20119
Name: count, dtype: int64

In [17]:
# Ticket price value patterns (top 15)
df['Ticket_Price_Cr'].value_counts().head(15)

Ticket_Price_Cr
100.0        278
70.0         271
60.0         268
₹100.0 Cr     73
₹70.0 Cr      64
₹60.0 Cr      56
₹11.71 Cr     41
₹12.63 Cr     36
₹11.04 Cr     35
₹10.48 Cr     35
₹11.56 Cr     34
₹10.37 Cr     34
₹12.0 Cr      34
₹12.24 Cr     33
₹10.65 Cr     33
Name: count, dtype: int64

In [18]:
# Unit size outliers
df[df['Unit_Size_Sqft'] <= 0]

Property_ID   Micro_Market Project_Name     Developer_Name  \
145     PROP000146   bellary road  Project_145        Puravankara   
238     PROP000239    KORAMANGALA  Project_238              Sobha   
244     PROP000245  sarjapur road  Project_244         L&T Realty   
289     PROP000290      Yelahanka  Project_289              Sobha   
747     PROP000748  sarjapur road  Project_247            Embassy   
...            ...            ...          ...                ...   
100069  PROP089998         domlur  Project_497             Godrej   
100087  PROP019615       jp nagar  Project_114         L&T Realty   
100553  PROP023080  Sarjapur Road   Project_79       Tata Housing   
100714  PROP078559         domlur   Project_58            Brigade   
100791  PROP086091      jayanagar   Project_90  Total Environment   

        Unit_Size_Sqft Configuration     Ticket_Price_Cr Transaction_Type  \
145               -1.0          3Bhk   9.094891145564691        Secondary   
238               -1.0         5Bhk+  11.065757877782282        Secondary   
244               -1.0         5BHK+   18.42659417975841        Secondary   
289               -1.0          4BHK  12.199936011524152        Secondary   
747               -1.0          3Bhk   9.448331344653205          Primary   
...                ...           ...                 ...              ...   
100069            -1.0         5BHK+   5.947726506609008        Secondary   
100087            -1.0         5BHK+  11.398934559342257        Secondary   
100553            -1.0          4Bhk  11.225770440381915        Secondary   
100714            -1.0          3Bhk  14.438398287100107          Primary   
100791            -1.0         5Bhk+  13.532688994916752        Secondary   

             Buyer_Type Purchase_Quarter  Connectivity_Score  Amenity_Score  \
145     Startup Founder       2024-03-31            9.952351       7.824742   
238               Other       2024-06-30            8.900902            NaN   
244     Startup Founder       2023-09-30            5.045976       7.885277   
289                 CXO       2024-06-30            7.194905       6.825103   
747               Other       2023-12-31            8.701167       5.982552   
...                 ...              ...                 ...            ...   
100069              HNI       2024-12-31            5.149795       8.012676   
100087            Other       2025-03-31            7.907058       9.525587   
100553              HNI       2023-09-30            7.973186       5.215596   
100714  Startup Founder       2025-03-31            9.061157       9.848616   
100791  Startup Founder       2024-12-31            8.147344       9.187069   

         Possession_Status Sales_Channel NRI_Buyer  Locality_Infra_Score  \
145          Ready to move        Direct       yes              6.690428   
238          Ready to move        Broker       yes              5.698885   
244     Under construction        Online       yes              5.766082   
289          Ready to move      NRI Desk        no              9.728299   
747          Ready to move        Online       yes              8.384907   
...                    ...           ...       ...                   ...   
100069  Under construction        Broker       yes              9.011854   
100087              Launch      NRI Desk        no              6.438825   
100553              Launch        Broker       yes              6.168787   
100714  Under construction      NRI Desk        no              8.069680   
100791  Under construction        Online       yes              8.181466   

        Avg_Traffic_Time_Min               Buyer_Comments  
145                       78         Loved the amenities!  
238                       43      Too far from my office.  
244                      102        Connectivity is poor.  
289                       21         Loved the amenities!  
747                       57         Loved the amenities!  
...                      ...   

In [19]:
df[df['Unit_Size_Sqft'] > 10000]

Empty DataFrame
Columns: [Property_ID, Micro_Market, Project_Name, Developer_Name, Unit_Size_Sqft, Configuration, Ticket_Price_Cr, Transaction_Type, Buyer_Type, Purchase_Quarter, Connectivity_Score, Amenity_Score, Possession_Status, Sales_Channel, NRI_Buyer, Locality_Infra_Score, Avg_Traffic_Time_Min, Buyer_Comments]
Index: []

In [20]:
# Purchase quarter distribution
df['Purchase_Quarter'].value_counts().sort_index()


Purchase_Quarter
2023-06-30    12655
2023-09-30    12618
2023-12-31    12560
2024-03-31    12750
2024-06-30    12608
2024-09-30    12683
2024-12-31    12595
2025-03-31    12531
Name: count, dtype: int64

### EDA Conclusions

Key data quality observations from raw dataset:

1. Missing Values:
   - Buyer_Comments: ~18%
   - Amenity_Score: ~10%
   - Unit_Size_Sqft: ~10%
   - Ticket_Price_Cr: ~10%

2. Data Type Issues:
   - Ticket_Price_Cr stored as text with currency symbols and mixed formats
   - Purchase_Quarter stored as string instead of datetime

3. Text Standardization Issues:
   - Micro_Market values contain casing and spelling variations
   - Configuration values are inconsistent (3Bhk, 3bhk, 3BHK, etc.)

4. Duplicate Records:
   - ~1000 duplicate Property_IDs detected

5. Invalid Numeric Values:
   - Unit_Size_Sqft contains negative values (-1)
   - Some unit sizes exceed realistic thresholds
